File: app/src/main/java/ch/threema/app/utils/WorkManagerUtil.java
/*  _____ _
 * |_   _| |_  _ _ ___ ___ _ __  __ _
 *   | | | ' \| '_/ -_) -_) '  \/ _` |_
 *   |_| |_||_|_| \___\___|_|_|_\__,_(_)
 *
 * Threema for Android
 * Copyright (c) 2014-2024 Threema GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package ch.threema.app.utils;

import com.google.common.util.concurrent.ListenableFuture;

import org.slf4j.Logger;

import java.util.List;
import java.util.Set;
import java.util.concurrent.ExecutionException;

import androidx.work.WorkInfo;
import androidx.work.WorkManager;
import ch.threema.base.utils.LoggingUtil;

public class WorkManagerUtil {
	private static final Logger logger = LoggingUtil.getThreemaLogger("WorkManagerUtil");

	/**
	 * Check if periodic work with provided tag is already scheduled or running and has the same schedule period.
	 * Cancel existing work in case of error
	 * @param workManager An instance of the WorkManager
	 * @param tag Unique work name
	 * @param schedulePeriod scheduled period of this work
	 * @return true if no periodic work with the same tag exists or the existing work has a different schedule period;
	 *      false if the work already exists and has the same schedule period
	 */
	public static boolean shouldScheduleNewWorkManagerInstance(WorkManager workManager, String tag, long schedulePeriod) {
		// check if work is already scheduled or running, if yes, do not attempt launch a new request
		ListenableFuture<List<WorkInfo>> workInfos = workManager.getWorkInfosForUniqueWork(tag);
		try {
			List<WorkInfo> workInfoList = workInfos.get();
			for (WorkInfo workInfo : workInfoList) {
				WorkInfo.State state = workInfo.getState();
				if (state == WorkInfo.State.RUNNING || state == WorkInfo.State.ENQUEUED) {
					logger.debug("a job of the same name is already running or queued");
					Set<String> tags = workInfo.getTags();
					if (tags.size() > 0 && tags.contains(String.valueOf(schedulePeriod))) {
						logger.debug("job has same schedule period");
						return false;
					} else {
						logger.debug("job has a different schedule period");
						break;
					}
				}
			}
		} catch (Exception e) {
			logger.info("WorkManager Exception");
			workManager.cancelUniqueWork(tag);
		}
		return true;
	}

	public static boolean isWorkManagerInstanceScheduled(WorkManager workManager, String tag) {
		ListenableFuture<List<WorkInfo>> workInfos = workManager.getWorkInfosForUniqueWork(tag);
		try {
			List<WorkInfo> workInfoList = workInfos.get();
			for (WorkInfo workInfo : workInfoList) {
				WorkInfo.State state = workInfo.getState();
				if (state == WorkInfo.State.ENQUEUED || state == WorkInfo.State.BLOCKED) {
					return true;
				}
			}
		} catch (ExecutionException | InterruptedException e) {
			logger.error("Could not get work info", e);
			return false;
		}
		return false;
	}

}

File: app/src/main/java/ch/threema/app/utils/WorkManagerUtil.kt
/*  _____ _
 * |_   _| |_  _ _ ___ ___ _ __  __ _
 *   | | | ' \| '_/ -_) -_) '  \/ _` |_
 *   |_| |_||_|_| \___\___|_|_|_\__,_(_)
 *
 * Threema for Android
 * Copyright (c) 2023-2024 Threema GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package ch.threema.app.utils

import android.content.Context
import android.text.format.DateUtils
import androidx.work.Operation
import androidx.work.WorkInfo
import androidx.work.WorkManager
import androidx.work.await
import ch.threema.base.utils.LoggingUtil
import java.util.concurrent.ExecutionException

private val logger = LoggingUtil.getThreemaLogger("WorkManagerUtil")

object WorkManagerUtil {
    @JvmStatic
    fun cancelUniqueWork(context: Context, uniqueWorkName: String): Operation {
        logger.info("Cancel unique work '{}'", uniqueWorkName)
        return WorkManager.getInstance(context).cancelUniqueWork(uniqueWorkName)
    }

    suspend fun cancelUniqueWorkAwait(context: Context, uniqueWorkName: String) {
        logger.info("Cancel result = {}", cancelUniqueWork(context, uniqueWorkName).await())
    }

    /**
     * Check if periodic work with provided [uniqueWorkName] is already scheduled or running and has the same schedule period.
     * Cancel existing work in case of error
     *
     * @param workManager An instance of the WorkManager
     * @param uniqueWorkName Unique work name
     * @param schedulePeriod scheduled period of this work
     * @return true if no periodic work with the same tag exists or the existing work has a different schedule period;
     *      false if the work already exists and has the same schedule period
     */
    @JvmStatic
    fun shouldScheduleNewWorkManagerInstance(workManager: WorkManager, uniqueWorkName: String, schedulePeriod: Long): Boolean {
        return try {
            workManager.getWorkInfosForUniqueWork(uniqueWorkName).get().none {
                val state = it.state
                if (state == WorkInfo.State.RUNNING || state == WorkInfo.State.ENQUEUED) {
                    logger.debug("A job of the same name is already running or queued")
                    if (it.tags.contains(schedulePeriod.toString())) {
                        logger.debug("Job has same schedule period")
                        true
                    } else {
                        logger.debug("Job has a different schedule period")
                        false
                    }
                } else {
                    false
                }
            }
        } catch (e: Exception) {
            logger.info("WorkManager Exception")
            workManager.cancelUniqueWork(uniqueWorkName)
            true
        }
    }

    @JvmStatic
    fun isWorkManagerInstanceScheduled(workManager: WorkManager, uniqueWorkName: String): Boolean {
        return try {
            workManager.getWorkInfosForUniqueWork(uniqueWorkName).get().any {
                val state = it.state
                state == WorkInfo.State.ENQUEUED || state == WorkInfo.State.BLOCKED
            }
        } catch (e: Exception) {
            when (e) {
                is ExecutionException, is InterruptedException -> logger.error("Could not get work info", e)
                else -> throw e
            }
            false
        }
    }

    /**
     * Normalize a schedule period in seconds to milliseconds:
     *
     * When [schedulePeriodS] is <= 0, a period of one day is returned.
     * Otherwise the [schedulePeriodS] is converted to milliseconds.
     *
     * @return The normalized schedule period in milliseconds
     */
    @JvmStatic
    fun normalizeSchedulePeriod(schedulePeriodS: Int): Long {
        return when {
            schedulePeriodS <= 0 -> DateUtils.DAY_IN_MILLIS
            else -> schedulePeriodS * DateUtils.SECOND_IN_MILLIS
        }
    }
}
